src/main/java/subway/domain/Sections.java

```diff
@@ -1,120 +1,152 @@
+            .collect(Collectors.toUnmodifiableSet());
     }
 
     public Sections() {
         this(new ArrayList<>());
     }
 
     public Sections addSection(final Section section) {
-        validateBothMatches(section);
-        validateNoMatches(section);
+        validateNotAlreadyExist(section);
 
         List<Section> newSections = new ArrayList<>(this.sections);
 
-        Section oldSection = findTargetSection(section);
-
-        if (oldSection != null) {
-            newSections.remove(oldSection);
-            newSections.add(oldSection.subtract(section));
-        }
-
+        findMatchingOneSection(section).ifPresent(oneMatchingStation -> {
+            newSections.remove(oneMatchingStation);
+            newSections.add(oneMatchingStation.subtract(section));
+        });
         newSections.add(section);
         return new Sections(newSections);
     }
 
-    private void validateBothMatches(final Section section) {
+    private void validateNotAlreadyExist(final Section section) {
         if (contains(section.getUpStation()) && contains(section.getDownStation())) {
-            throw new IllegalArgumentException("두 역 모두 기존 노선에 포함될 수 없습니다.");
+            throw new SubwayException(INVALID_SECTION_ALREADY_EXISTS);
         }
-    }
-
-    private void validateNoMatches(final Section section) {
-        if (!contains(section.getUpStation()) && !contains(section.getDownStation())) {
-            throw new IllegalArgumentException("두 역 중 하나는 기존 노선에 포함되어야 합니다");
+        if (notContains(section.getUpStation()) && notContains(section.getDownStation())) {
+            throw new SubwayException(INVALID_SECTION_NO_EXISTS);
         }
     }
 
     private boolean contains(final Station station) {
-        return sections.stream()
-                .flatMap(section -> Stream.of(section.getUpStation(), section.getDownStation()))
-                .distinct()
-                .collect(Collectors.toList())
-                .contains(station);
+        return cache.contains(station);
+    }
+
+    private boolean notContains(final Station station) {
+        return !contains(station);
     }
 
-    private Section findTargetSection(final Section section) {
+    private Optional<Section> findMatchingOneSection(final Section section) {
         return sections.stream()
-                .filter(section::matchOneStation)
-                .findAny()
-                .orElse(null);
+            .filter(section::matchOneStation)
+            .findAny();
     }
 
     public Sections removeStation(final Station station) {
-        validateDownStationTerminal(station);
         validateSize();
+        validateContainStation(station);
+        List<Section> newSections = new ArrayList<>(this.sections);
+        List<Section> removeSection = findRemoveSections(station, newSections);
+
+        newSections.removeAll(removeSection);
+
+        if (isMiddleStation(station)) {
+            newSections.add(reArrangeSection(station));
```

Comment:
```suggestion
            newSections.add(rearrangeSection(station));
```

하나의 단어라서 중간에 camelCase 하지 않아도 됩니다 😉

Reply:
넵 수정하겠습니다!